### ISSUE TITLE
examples/testscript.py hangs on Kmeans clustering

### ISSUE THREAD

**Issue by tbenst** [DeepLabCut/DeepLabCut#637]
**Describe the bug**
For deeplabcut v2.1.6, I am getting a hang when running `examples/testscript.py`. I have seen `Kmeans clustering ... (this might take a while)` for over an hour now, on an 8-core Intel i7-9800X X-series processor. All 16 threads are slammed at 100% utilization. I'm happy to keep waiting, but the script comments say this script should finish in 3:30 minutes (which I read as 3min 30sec?) on CPU...unless that is 3 hours 30 minutes?

**To Reproduce**
```
# install nix if not already installed
> curl -L https://nixos.org/nix/install | sh
> git clone https://github.com/tbenst/nixpkgs.git
> cd nixpkgs
> git checkout deeplabcut
> nix-build -I nixpkgs=. -A python3Packages.deeplabcut
```

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information about your system):**
 - OS: **NixOS 20.03**
 - DeepLabCut Version [e.g. 22] (please check with ``import deeplabcut``, ``deeplabcut.__version__``) **2.1.6**

**Additional context**
I have slightly modified deeplabcut's required python packages to use Nix's distributed versions: https://github.com/tbenst/nixpkgs/blob/ced8ca5dba8f2e00cd45fd5904f0b08593db205a/pkgs/development/python-modules/deeplabcut/default.nix#L103. Additionally, I run testscript headlessly using `xvfb-run`.

**Comment by AlexEMG** [comment]
Sure, if that is for our example data than that takes way too long. So, I suppose that's something about your context? Can you try to figure out where in the code it is stuck? I.e. does it load the video etc.

**Comment by MMathisLab** [comment]
I don't know much about nix packages, but perhaps the lack of  intel0openmp is an issue?

 ``    --replace "'intel-openmp'," "" \ ``

**Comment by tbenst** [comment]
@MMathisLab thanks, I added that package (binaries should already be there as nix packages intel-openmy with the binary MKL package) just in case, no difference.

I used `pudb` to track down the hang all the way until [_k_init](https://github.com/scikit-learn/scikit-learn/blob/a84ba0c0b8b4948a5571ce1a815102c43d35cdd1/sklearn/cluster/_kmeans.py#L120). This calls [euclidean_distances](https://github.com/scikit-learn/scikit-learn/blob/a84ba0c0b8b4948a5571ce1a815102c43d35cdd1/sklearn/metrics/pairwise.py#L303), where the code calls [safe_sparse_dot](https://github.com/scikit-learn/scikit-learn/blob/1495f69242646d239d89a5713982946b8ffcf9d9/sklearn/utils/extmath.py#L142). The hang occurs in the `np.dot` call, and I couldn't trace further with a python debugger. I did save the two arrays being sent to `np.dot` (`a.shape == (3, 810)`, `b.shape ==(810, 256)`). but I could not reproduce the hang in a python shell. Therefore, we appear to have hit some ugly calling context-dependency in the compiled numpy code, which is a major bummer.

I'm on sklearn 0.21.3, and I noticed that `safe_sparse_dot` changes dramatically. Now, rather than calling `np.dot`, the code calls `a @ b` unless ndim > 2. Thus, perhaps we hit a known bug that might be fixed on upgrading sklearn. 

In fact, this appears to be the same issue as described back in 2012: https://github.com/scikit-learn/scikit-learn/issues/636#issuecomment-4172159

Edit: worth noting that I’m on Linux using MKL. Explanation appears to be https://github.com/scikit-learn/scikit-learn/issues/636#issuecomment-134504345 and may be related to Python Multiprocessing violating POSIX standard. In any case, this does not appear to be a deeplabcut problem

Edit 2: I was able to fix the hang by changing some dependencies! It appears that the issue only occurs if `numpy` is compiled against `openblas`, and does not occur if `numpy` is compiled against `mkl`. I am working to validate this today

**Comment by tbenst** [comment]
Ok, much progress! The default cases without Tensorpack loader are functional: [output](https://gist.github.com/tbenst/db8f258c7f3bdc0f8d1d024342513798).

Seems like this stalls at `deeplabcut.train_network(path_config_file,shuffle=2,allow_growth=True)`. There is a warning that may be helpful:
```
[0331 19:38:05 @parallel.py:355] [MultiProcessRunnerZMQ] Will fork a dataflow more than one times. This assumes the datapoints are i.i.d.
[0331 19:38:06 @argtools.py:146] WRN Starting a process with 'fork' method is not safe and may consume unnecessary extra CPU memory. Use 'forkserver/spawn' method (available after Py3.4) instead if you run into any issues. See https://docs.python.org/3/library/multiprocessing.html#contexts-and-start-methods
```

Any ideas? Many thanks

**Comment by AlexEMG** [comment]
I think that process is from tensorpack I think.

________________________________

**Comment by MMathisLab** [comment]
Yes, this is from tensorpack, if you don't run on a linux (ubuntu, etc) this is a warning one gets/the process never stops... hence why we test for defualt loader, imgaug, and then only tensorpack support for ubuntu :) although interesting you get this on MKL

**Comment by AlexEMG** [comment]
https://github.com/AlexEMG/DeepLabCut/blob/master/examples/testscript.py#L41

The latest testscript excludes tensorpack though.

________________________________

**Comment by tbenst** [comment]
Yes, this is being run on Linux (although NixOS is admittedly a less common distribution) & I'm running the latest DLC tag. I was not successful in running the tensorpack tests, so I agree with your diagnosis that tensorpack is the issue, thank you!

Btw, I can confirm that the KMeans hang is indeed solely due to `openblas`, I ran `testscript.py` with all the same dependencies that worked with `MKL` so conclusively isolated that problem at least.

**Comment by MMathisLab** [comment]
I will close the issue, since we know the culprit regarding openblas, but please feel free to post more if something comes up!
